go/engine/identify2.go

```diff
@@ -0,0 +1,486 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
+
+import (
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+	"sync"
+	"time"
+)
+
+var locktab libkb.LockTable
+
+type identify2TestArgs struct {
+	noMe     bool                  // don't load ME
+	tcl      *libkb.TrackChainLink // the track chainlink to use
+	selfLoad bool                  // on if this is a self load
+	cache    libkb.Identify2Cacher
+	clock    func() time.Time
+}
+
+//
+// TODOs:
+//   - think harder about what we're caching in failure cases; right now we're only
+//     caching full successes.
+//   - Better error typing for various failures.
+//   - Work back in the identify card
+//
+
+// Identify2 is the Identify engine used in KBFS and as a subroutine
+// of command-line crypto.
+type Identify2 struct {
+	libkb.Contextified
+
+	arg        *keybase1.Identify2Arg
+	testArgs   *identify2TestArgs
+	trackToken keybase1.TrackToken
+	cachedRes  *keybase1.UserPlusKeys
+
+	me   *libkb.User
+	them *libkb.User
+
+	themAssertion   libkb.AssertionExpression
+	remoteAssertion libkb.AssertionAnd
+	localAssertion  libkb.AssertionAnd
+
+	state        libkb.IdentifyState
+	useTracking  bool
+	identifyKeys []keybase1.IdentifyKey
+}
+
+var _ (Engine) = (*Identify2)(nil)
+
+// Name is the unique engine name.
+func (e *Identify2) Name() string {
+	return "Identify2"
+}
+
+func NewIdentify2(g *libkb.GlobalContext, arg *keybase1.Identify2Arg) *Identify2 {
+	return &Identify2{
+		Contextified: libkb.NewContextified(g),
+		arg:          arg,
+	}
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *Identify2) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// Run then engine
+func (e *Identify2) Run(ctx *Context) (err error) {
+
+	e.G().Log.Debug("+ Identify2.runSingle(UID=%v, Assertion=%s)", e.arg.Uid, e.arg.UserAssertion)
+
+	e.G().Log.Debug("+ acquire singleflight lock")
+	lock := locktab.AcquireOnName(e.arg.Uid.String())
```

Comment:
this can be made more robust if it worked more like a WaitGroup and for every scope we go into, we `Add(1)` and `defer Done()` on it.